secret-rotator/client/client.go

```diff
@@ -0,0 +1,289 @@
+		Payload: &secretmanagerpb.SecretPayload{
+			Data: data,
+		},
+	}
+	verResp, err := cl.AddSecretVersion(context.TODO(), verReq)
+	if err != nil {
+		return "", err
+	}
+
+	// extract the latest version number instead of storing the entire verResp.Name
+	// because '/'s are not allowed in gsm metedata
+	splits := strings.Split(verResp.Name, "/")
+	latestVersion := splits[len(splits)-1]
+
+	return latestVersion, nil
+}
+
+// GetCreateTime gets the createTime of the secret version specified by project, id, version.
+// Returns createTime if successful, otherwise error.
+func (cl *Client) GetCreateTime(project, id, version string) (time.Time, error) {
+	ctx := context.TODO()
+	name := "projects/" + project + "/secrets/" + id + "/versions/" + version
+
+	getReq := &secretmanagerpb.GetSecretVersionRequest{
+		Name: name,
+	}
+	getResult, err := cl.GetSecretVersion(ctx, getReq)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	createTime, err := ptypes.Timestamp(getResult.CreateTime)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	return createTime, nil
+}
+
+// GetSecretLabels gets the labels of the secret specified by project, id.
+// Returns secret labels if successful, otherwise error
+func (cl *Client) GetSecretLabels(project, id string) (map[string]string, error) {
+	ctx := context.TODO()
+	name := "projects/" + project + "/secrets/" + id
+
+	getReq := &secretmanagerpb.GetSecretRequest{
+		Name: name,
+	}
+	getResult, err := cl.GetSecret(ctx, getReq)
+	if err != nil {
+		return nil, err
+	}
+
+	return getResult.Labels, nil
+}
+
+// GetSecretVersionData gets the data of the secret version specified by project, id, version.
+// Returns secret value if successful, otherwise error
+func (cl *Client) GetSecretVersionData(project, id, version string) ([]byte, error) {
+	ctx := context.TODO()
+	name := "projects/" + project + "/secrets/" + id + "/versions/" + version
+
+	accReq := &secretmanagerpb.AccessSecretVersionRequest{
+		Name: name,
+	}
+	accResult, err := cl.Client.AccessSecretVersion(ctx, accReq)
+	if err != nil {
+		return nil, err
+	}
+
+	return accResult.Payload.Data, nil
+}
+
+// GetSecretVersionState gets the state of the secret version specified by project, id, version.
+// Returns state if successful, otherwise error.
+func (cl *Client) GetSecretVersionState(project, id, version string) (int32, error) {
+	ctx := context.TODO()
+	name := "projects/" + project + "/secrets/" + id + "/versions/" + version
+
+	getReq := &secretmanagerpb.GetSecretVersionRequest{
+		Name: name,
+	}
+	getResult, err := cl.GetSecretVersion(ctx, getReq)
+
+	return int32(getResult.State), err
+}
+
+// ChangeSecretVersionState changes the state of secret version
+// returns nil if successful, otherwise error.
+func (cl *Client) ChangeSecretVersionState(project, id, version string, state int32) error {
+	ctx := context.TODO()
+	name := "projects/" + project + "/secrets/" + id + "/versions/" + version
+
+	switch state {
+	case 1: // SecretVersion_ENABLED
```

Comment:
Yes, it's defined in https://github.com/googleapis/go-genproto/blob/da3ae01ef02dd0f1909d902122c432a450c6efb3/googleapis/cloud/secretmanager/v1/resources.pb.go#L46

Reply:
I was imported as alias `secretmanagerpb`, and I thought it's too lengthy to write `secretmanagerpb.SecretVersion_ENABLED` everywhere.

But thiking again now, maybe I could just make the alias shorter and use something like `gsmpb.SecretVersion_ENABLED` .

Reply:
> This feels like an unfriendly API, I guess I'm curious why we didn't make similar `EnableSecretVersion` etc. methods

Because I was trying to avoid the function naming of `client.EnableSecretVersion()`.
I have the `secretmanager.Client` embedded in `rotator.Client`, like 
```
type Client struct {
	*secretmanager.Client
}
``` 
so previously it always gave me errors if I had identical function names with `secretmanager.Client`.

But this could actually be easily solved, just call `cl.Client.EnableSecretVersion` within a `rotator.Client` method.
So I think I'll break this into `EnableSecretVersion`, `DisableSecretVersion`, and `DestroySecretVersion`.